File: alumnium/alumni.py
from appium.webdriver.webdriver import WebDriver as Appium
from playwright.sync_api import Page
from retry import retry
from selenium.webdriver.remote.webdriver import WebDriver

from alumnium.client import Client
from alumnium.tools import ALL_APPIUM_TOOLS, ALL_TOOLS, BaseTool

from .agents import *
from .agents.retriever_agent import Data
from .area import Area
from .drivers import AppiumDriver, PlaywrightDriver, SeleniumDriver
from .logutils import get_logger
from .models import Model

logger = get_logger(__name__)


class Alumni:
    def __init__(self, driver: Page | WebDriver, model: Model = None):
        self.model = model or Model.current

        if isinstance(driver, Appium):
            self.driver = AppiumDriver(driver)
            self.tools = ALL_APPIUM_TOOLS
        elif isinstance(driver, Page):
            self.driver = PlaywrightDriver(driver)
            self.tools = ALL_TOOLS
        elif isinstance(driver, WebDriver):
            self.driver = SeleniumDriver(driver)
            self.tools = ALL_TOOLS
        else:
            raise NotImplementedError(f"Driver {driver} not implemented")

        logger.info(f"Using model: {self.model.provider.value}/{self.model.name}")

        self.client = Client(self.model, self.tools)
        self.cache = self.client.cache

    def quit(self):
        self.client.quit()
        self.driver.quit()

    @retry(tries=2, delay=0.1)
    def do(self, goal: str):
        """
        Executes a series of steps to achieve the given goal.

        Args:
            goal: The goal to be achieved.
        """
        initial_accessibility_tree = self.driver.accessibility_tree
        steps = self.client.planner_agent.invoke(goal, initial_accessibility_tree.to_xml())
        for idx, step in enumerate(steps):
            # If the step is the first step, use the initial accessibility tree.
            accessibility_tree = initial_accessibility_tree if idx == 0 else self.driver.accessibility_tree
            actor_response = self.client.actor_agent.invoke(goal, step, accessibility_tree.to_xml())

            # Execute tool calls
            for tool_call in actor_response:
                BaseTool.execute_tool_call(tool_call, self.tools, accessibility_tree, self.driver)

    def check(self, statement: str, vision: bool = False) -> str:
        """
        Checks a given statement true or false.

        Args:
            statement: The statement to be checked.
            vision: A flag indicating whether to use a vision-based verification via a screenshot. Defaults to False.

        Returns:
            The summary of verification result.

        Raises:
            AssertionError: If the verification fails.
        """
        result = self.client.retriever_agent.invoke(
            f"Is the following true or false - {statement}",
            self.driver.accessibility_tree.to_xml(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
        )
        assert result[1], result[0]
        return result[0]

    def get(self, data: str, vision: bool = False) -> Data:
        """
        Extracts requested data from the page.

        Args:
            data: The data to extract.
            vision: A flag indicating whether to use a vision-based extraction via a screenshot. Defaults to False.

        Returns:
            Data: The extracted data loosely typed to int, float, str, or list of them.
        """
        return self.client.retriever_agent.invoke(
            data,
            self.driver.accessibility_tree.to_xml(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
        )[1]

    def area(self, description: str) -> Area:
        """
        Creates an area for the agents to work within.
        This is useful for narrowing down the context or focus of the agents' actions, checks and data retrievals.

        Note that if the area cannot be found, the topmost area of the accessibility tree will be used,
        which is equivalent to the whole page.

        Args:
            description: The description of the area.

        Returns:
            Area: An instance of the Area class that represents the area of the accessibility tree to use.
        """
        response = self.client.area_agent.invoke(description, self.driver.accessibility_tree.to_xml())
        return Area(
            id=response["id"],
            description=response["explanation"],
            driver=self.driver,
            tools=self.tools,
            client=self.client,
        )

    def learn(self, goal: str, actions: list[str]):
        """
        Adds a new learning example on what steps should be take to achieve the goal.

        Args:
            goal: The goal to be achieved. Use same format as in `do`.
            actions: A list of actions to achieve the goal.
        """
        self.client.planner_agent.add_example(goal, actions)

    def clear_learn_examples(self):
        """
        Clears the learn examples.
        """
        self.client.planner_agent.prompt_with_examples.examples.clear()

    def stats(self) -> dict[str, int]:
        """
        Returns the stats of the session.
        """
        return self.client.session.stats()

File: packages/typescript/src/Alumni.ts
import { Page } from "playwright";
import { WebDriver } from "selenium-webdriver";
import type { Browser } from "webdriverio";
import { Area } from "./Area.js";
import { Cache } from "./Cache.js";
import { HttpClient } from "./clients/HttpClient.js";
import { Data } from "./clients/typecasting.js";
import { AppiumDriver } from "./drivers/AppiumDriver.js";
import { BaseDriver } from "./drivers/BaseDriver.js";
import { Element } from "./drivers/index.js";
import { PlaywrightDriver } from "./drivers/PlaywrightDriver.js";
import { SeleniumDriver } from "./drivers/SeleniumDriver.js";
import { Model } from "./Model.js";
import { BaseTool, ToolCall } from "./tools/BaseTool.js";
import { ClickTool } from "./tools/ClickTool.js";
import { DragAndDropTool } from "./tools/DragAndDropTool.js";
import { HoverTool } from "./tools/HoverTool.js";
import { PressKeyTool } from "./tools/PressKeyTool.js";
import { SelectTool } from "./tools/SelectTool.js";
import { TypeTool } from "./tools/TypeTool.js";
import { getLogger } from "./utils/logger.js";

const logger = getLogger(["Alumni"]);

export interface AlumniOptions {
  url?: string;
  model?: Model;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extraTools?: (new (...args: any[]) => BaseTool)[];
}

export interface CheckOptions {
  vision?: boolean;
}

export interface GetOptions {
  vision?: boolean;
}

export class Alumni {
  public driver: BaseDriver;
  private client: HttpClient;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private tools: Record<string, new (...args: any[]) => BaseTool>;
  public cache: Cache;
  private url: string;
  private model: Model;

  constructor(driver: WebDriver | Page | Browser, options: AlumniOptions = {}) {
    this.url = options.url || "http://localhost:8013";
    this.model = options.model || Model.current;

    // Wrap driver or use directly if already wrapped
    if (driver instanceof WebDriver) {
      this.driver = new SeleniumDriver(driver);
    } else if ((driver as Page).context) {
      this.driver = new PlaywrightDriver(driver as Page);
    } else if (
      (driver as Browser).capabilities &&
      (driver as Browser).getPageSource
    ) {
      // WebdriverIO Browser (Appium)
      this.driver = new AppiumDriver(driver as Browser);
    } else {
      throw new Error(`Unsupported driver type '${typeof driver}'`);
    }

    // Initialize tools
    this.tools = {
      ClickTool,
      TypeTool,
      HoverTool,
      SelectTool,
      PressKeyTool,
      DragAndDropTool,
    };

    // Add extra tools if provided
    if (options.extraTools) {
      for (const tool of options.extraTools) {
        this.tools[tool.name] = tool;
      }
    }

    // Initialize HTTP client
    this.client = new HttpClient(this.url, this.driver.platform, this.tools);
    this.cache = new Cache(this.client);

    logger.info(`Using model: ${this.model.provider}/${this.model.name}`);
    logger.info(`Using HTTP client with server: ${this.url}`);
  }

  async quit(): Promise<void> {
    await this.client.quit();
    await this.driver.quit();
  }

  async do(goal: string): Promise<void> {
    const initialAccessibilityTree = await this.driver.getAccessibilityTree();
    const steps = await this.client.planActions(
      goal,
      initialAccessibilityTree.toStr()
    );

    for (let idx = 0; idx < steps.length; idx++) {
      const step = steps[idx];

      // Use initial tree for first step, fresh tree for subsequent steps
      const accessibilityTree =
        idx === 0
          ? initialAccessibilityTree
          : await this.driver.getAccessibilityTree();
      const actorResponse = await this.client.executeAction(
        goal,
        step,
        accessibilityTree.toStr()
      );

      // Execute tool calls
      for (const toolCall of actorResponse) {
        await BaseTool.executeToolCall(
          toolCall as ToolCall,
          this.tools,
          this.driver
        );
      }
    }
  }

  async check(statement: string, options: CheckOptions = {}): Promise<string> {
    const screenshot = options.vision
      ? await this.driver.screenshot()
      : undefined;
    const accessibilityTree = await this.driver.getAccessibilityTree();
    const [explanation, value] = await this.client.retrieve(
      `Is the following true or false - ${statement}`,
      accessibilityTree.toStr(),
      await this.driver.title(),
      await this.driver.url(),
      screenshot
    );

    if (!value) {
      throw new Error(explanation);
    }

    return explanation;
  }

  async get(data: string, options: GetOptions = {}): Promise<Data> {
    const screenshot = options.vision
      ? await this.driver.screenshot()
      : undefined;
    const accessibilityTree = await this.driver.getAccessibilityTree();
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [_explanation, value] = await this.client.retrieve(
      data,
      accessibilityTree.toStr(),
      await this.driver.title(),
      await this.driver.url(),
      screenshot
    );

    return value;
  }

  async find(description: string): Promise<Element> {
    const accessibilityTree = await this.driver.getAccessibilityTree();
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const response = await this.client.findElement(
      description,
      accessibilityTree.toStr()
    );
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    return this.driver.findElement(response.id as number);
  }

  async area(description: string): Promise<Area> {
    const accessibilityTree = await this.driver.getAccessibilityTree();
    const response = await this.client.findArea(
      description,
      accessibilityTree.toStr()
    );
    const scopedTree = accessibilityTree.scopeToArea(response.id);
    return new Area(
      response.id,
      response.explanation,
      scopedTree,
      this.driver,
      this.tools,
      this.client
    );
  }

  async learn(goal: string, actions: string[]): Promise<void> {
    await this.client.addExample(goal, actions);
  }

  async clearLearnExamples(): Promise<void> {
    await this.client.clearExamples();
  }

  async getStats(): Promise<Record<string, Record<string, number>>> {
    return await this.client.getStats();
  }
}
